README.md

```diff
@@ -17,15 +17,19 @@ repeatable way, e.g. in the context of deployment tools:
 hosts = ["1.eg.io", {"2.eg.io", port: 2222}]
 
 context =
-  SSHKit.context(hosts)
+  SSHKit.context()
   |> SSHKit.path("/var/www/phx")
   |> SSHKit.user("deploy")
   |> SSHKit.group("deploy")
   |> SSHKit.umask("022")
   |> SSHKit.env(%{"NODE_ENV" => "production"})
 
-:ok = SSHKit.upload(context, ".", recursive: true)
-:ok = SSHKit.run(context, "yarn install")
+{:ok, conn} = SSHKit.connect(hosts)
+
+{:ok, conn} = SSHKit.upload(conn, context, ".", recursive: true)
```

Comment:
I'm missing the overview if this hinders some current usecases, but in my personal opinion, the API would be cleaner and easier to jump in if both, `upload` and `run` would **not** take a `context`, but have all they they need already in the `conn` that is passed it.

While this means, `connect` needs to be called sometime first in order to get a "connection", I think this pipeline will actually feel super natural for anyone using the library.

Reply:
Awesome @klappradla. Thanks a ton for sharing your thoughts! 💚 

My idea was to explicitly separate `conn` and `context`. My current (and by no means only) way of thinking about these is:

1. You establish an SSH **connection** with a remote (basically `ssh user@remote`).
2. You perform several operations via this connection. Each operation executes within a certain **context** (path, env…). Several operations executed via the same connection may use different contexts.
3. You close the connection.

That's why I came to the conclusion that keeping them separate might be a good idea. If we decide to keep both in a single struct, we'd tie the execution context to a specific connection and vice versa.

Can you elaborate a bit more on why you would like to keep them in a single struct? I would love to look at a few more alternatives. Writing them down probably helps 🙂 

I imagine a combined struct variant would look something like this:

```elixir
hosts = […]

{:ok, conn} =
  hosts
  |> SSHKit.connect()
  |> SSHKit.path(…)
  |> SSHKit.user(…)

{:ok, conn} = SSHKit.run(conn, "mix deps.get")

{:ok, _} = SSHKit.close(conn)
```

The downside of this approach seems to me, that it's not as straight-forward to run multiple commands via the same connection in different contexts (e.g. at different paths…). But maybe I am misreading your comment. 😊 

Happy to hear about your ideas 🙇

Reply:
💡 thanx for the additional explanation 🙏. I did not understand the purpose of both well enought 🙈 . With my increased knowledge, your idea of separating them actually makes a ton of sense and will probably lead to a cleaner and easier to understand API of the whole library.

The way I understood "reusing" the connection with multiple contexts initially was: the `conn` includes the `context` and one can build a new one (with a new context) with passing in an existing `conn` and without actually establishing a new "connection".
A bit like *"mutating the context of a connection*" 😉

Reply:
@klappradla what you had in mind would also work. I imagine it'd look something like this:

```elixir
{:ok, conn} = SSHKit.connect(hosts)

# Run a command without a specific context.
{:ok, conn, output} = SSHKit.run(conn, "whoami")

# Now we run another command through the same connection but as a different user:
{:ok, conn, output2} =
  conn
  |> SSHKit.user("alter_ego")
  |> SSHKit.run("whoami")

# What should conn be now? Does it remember the new user?
```

The last question is why I was thinking to keep both separate. Maybe we can stick our heads together tomorrow, to see whether I really got your idea 😀